src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -172,190 +185,132 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
     {
         while (!reader.End)
         {
+            // Size of the detected EOL
+            var eolSize = -1;
+
+            // Check if the reader's span contains an LF to skip the reader if possible
             var span = reader.UnreadSpan;
-            while (span.Length > 0)
+
+            // Fast path, CR/LF at the beginning
+            if (span.Length >= 2 && span[0] == ByteCR && span[1] == ByteLF)
             {
-                byte ch1;
-                var ch2 = (byte)0;
-                var readAhead = 0;
+                reader.Advance(2);
+                handler.OnHeadersComplete(endStream: false);
+                return true;
+            }
 
-                // Fast path, we're still looking at the same span
-                if (span.Length >= 2)
-                {
-                    ch1 = span[0];
-                    ch2 = span[1];
-                }
-                else if (reader.TryRead(out ch1)) // Possibly split across spans
-                {
-                    // Note if we read ahead by 1 or 2 bytes
-                    readAhead = (reader.TryRead(out ch2)) ? 2 : 1;
-                }
+            var lfIndex = span.IndexOf(ByteLF);
 
-                if (ch1 == ByteCR)
-                {
-                    // Check for final CRLF.
-                    if (ch2 == ByteLF)
-                    {
-                        // If we got 2 bytes from the span directly so skip ahead 2 so that
-                        // the reader's state matches what we expect
-                        if (readAhead == 0)
-                        {
-                            reader.Advance(2);
-                        }
-
-                        // Double CRLF found, so end of headers.
-                        handler.OnHeadersComplete(endStream: false);
-                        return true;
-                    }
-                    else if (readAhead == 1)
-                    {
-                        // Didn't read 2 bytes, reset the reader so we don't consume anything
-                        reader.Rewind(1);
-                        return false;
-                    }
+            if (lfIndex > -1)
+            {
+                span = span[..lfIndex];
```

Comment:
```suggestion
                span = span.Slice(0, lfIndex);
```
Right now Roslyn emits too much code for the range. Explicit slice has better codegen (for IL and machine code).
There's an issue open for Roslyn for quite some time, so I don't know when this will be fixed.